File: tests/test_reports_e2e.mjs
// Playwright E2E test for the redesigned Report system
import { chromium } from 'playwright';
import { writeFileSync } from 'fs';

const BASE = 'http://localhost:3000';
const SCREENSHOT_DIR = '/tmp/report-tests';

async function screenshot(page, name) {
  const path = `${SCREENSHOT_DIR}/${name}.png`;
  await page.screenshot({ path, fullPage: false });
  console.log(`  [screenshot] ${name}`);
  return path;
}

async function main() {
  // Ensure screenshot dir exists
  const { mkdirSync } = await import('fs');
  mkdirSync(SCREENSHOT_DIR, { recursive: true });

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 },
    colorScheme: 'dark',
  });
  const page = await context.newPage();

  // Increase default timeout
  page.setDefaultTimeout(15000);

  try {
    // ── Step 1: Login ──
    console.log('\n=== Step 1: Login ===');
    await page.goto(`${BASE}/login`);
    await page.waitForLoadState('networkidle');
    await screenshot(page, '01-login-page');

    // Fill login form — look for email/password inputs
    const emailInput = page.locator('input[type="email"], input[name="email"], input[placeholder*="mail"]').first();
    const passInput = page.locator('input[type="password"]').first();

    if (await emailInput.isVisible()) {
      await emailInput.fill('robert@investmentx.ai');
      await passInput.fill('admin');
      await screenshot(page, '02-login-filled');

      // Submit
      const submitBtn = page.locator('button[type="submit"], button:has-text("Sign In"), button:has-text("Login"), button:has-text("Log in")').first();
      await submitBtn.click();
      await page.waitForLoadState('networkidle');
      await page.waitForTimeout(2000);
      await screenshot(page, '03-after-login');
    } else {
      console.log('  No email input found, may already be logged in');
    }

    // ── Step 2: Navigate to Reports ──
    console.log('\n=== Step 2: Navigate to Reports ===');
    await page.goto(`${BASE}/reports`);
    await page.waitForLoadState('networkidle');
    await page.waitForTimeout(2000);
    await screenshot(page, '04-reports-list');

    // ── Step 3: Click New Report → Template Picker ──
    console.log('\n=== Step 3: Template Picker ===');
    const newBtn = page.locator('button:has-text("New Report")').first();
    if (await newBtn.isVisible()) {
      await newBtn.click();
      await page.waitForTimeout(1000);
      await screenshot(page, '05-template-picker');

      // Select "Macro Outlook" template
      const macroBtn = page.locator('button:has-text("Macro Outlook")').first();
      if (await macroBtn.isVisible()) {
        await macroBtn.click();
        console.log('  Selected Macro Outlook template');
      } else {
        // Fallback: click first template
        const firstTemplate = page.locator('button:has-text("Blank")').first();
        if (await firstTemplate.isVisible()) {
          await firstTemplate.click();
          console.log('  Selected Blank template');
        }
      }
      await page.waitForLoadState('networkidle');
      await page.waitForTimeout(2000);
      await screenshot(page, '06-report-editor');
    } else {
      console.log('  New Report button not found');
      await screenshot(page, '05-no-new-button');
    }

    // ── Step 4: Editor interaction ──
    console.log('\n=== Step 4: Editor Interaction ===');

    // Check if we're in the editor (should see slide thumbnails)
    const editor = page.locator('[class*="aspect-[16/9]"]').first();
    if (await editor.isVisible({ timeout: 5000 }).catch(() => false)) {
      console.log('  Editor canvas visible');

      // Edit the title
      const titleInput = page.locator('input[placeholder*="Title"], input[placeholder*="title"]').first();
      if (await titleInput.isVisible({ timeout: 3000 }).catch(() => false)) {
        await titleInput.fill('Q2 2026 Macro Outlook');
        console.log('  Filled title');
      }
      await screenshot(page, '07-editor-with-title');

      // Click on slide 2 (if exists)
      const slide2 = page.locator('[class*="rounded-"][class*="border"][class*="cursor-pointer"]').nth(1);
      if (await slide2.isVisible({ timeout: 3000 }).catch(() => false)) {
        await slide2.click();
        await page.waitForTimeout(500);
        await screenshot(page, '08-slide-2');
        console.log('  Switched to slide 2');
      }

      // Try adding a new slide via "+ New Slide" button
      const newSlideBtn = page.locator('button:has-text("New Slide")').first();
      if (await newSlideBtn.isVisible({ timeout: 3000 }).catch(() => false)) {
        await newSlideBtn.click();
        await page.waitForTimeout(1000);
        await screenshot(page, '09-layout-picker');
        console.log('  Layout picker opened');

        // Select chart_full layout
        const chartFullBtn = page.locator('button:has-text("Chart")').first();
        if (await chartFullBtn.isVisible({ timeout: 3000 }).catch(() => false)) {
          await chartFullBtn.click();
          await page.waitForTimeout(500);
          await screenshot(page, '10-new-chart-slide');
          console.log('  Added chart slide');
        }
      }

      // Try opening chart picker by clicking the "Click to add chart" area
      const addChartBtn = page.locator('button:has-text("Click to add chart")').first();
      if (await addChartBtn.isVisible({ timeout: 3000 }).catch(() => false)) {
        await addChartBtn.click();
        await page.waitForTimeout(1500);
        await screenshot(page, '11-chart-picker');
        console.log('  Chart picker opened');

        // Check if charts are listed
        const chartList = page.locator('button:has(svg.lucide-line-chart)');
        const chartCount = await chartList.count();
        console.log(`  Found ${chartCount} charts in picker`);

        if (chartCount > 0) {
          // Click first chart to preview
          await chartList.first().click();
          await page.waitForTimeout(2000);
          await screenshot(page, '12-chart-preview');
          console.log('  Chart preview shown');

          // Click "Insert Chart" button
          const insertBtn = page.locator('button:has-text("Insert Chart")').first();
          if (await insertBtn.isVisible({ timeout: 3000 }).catch(() => false)) {
            await insertBtn.click();
            await page.waitForTimeout(2000);
            await screenshot(page, '13-chart-inserted');
            console.log('  Chart inserted into slide');
          }
        }

        // Close picker if still open
        const closeBtn = page.locator('button[aria-label="Close"]').first();
        if (await closeBtn.isVisible({ timeout: 1000 }).catch(() => false)) {
          await closeBtn.click();
        }
      }

      // ── Step 5: Test duplicate slide ──
      console.log('\n=== Step 5: Duplicate & Properties ===');
      const duplicateBtn = page.locator('button[title*="Duplicate"]').first();
      if (await duplicateBtn.isVisible({ timeout: 2000 }).catch(() => false)) {
        await duplicateBtn.click();
        await page.waitForTimeout(500);
        console.log('  Slide duplicated');
      }

      // Open properties panel
      const propsBtn = page.locator('button[title*="properties"]').first();
      if (await propsBtn.isVisible({ timeout: 2000 }).catch(() => false)) {
        await propsBtn.click();
        await page.waitForTimeout(500);
        await screenshot(page, '14-properties-panel');
        console.log('  Properties panel opened');
      }

      // ── Step 6: Test layout switcher ──
      console.log('\n=== Step 6: Layout Switcher ===');
      await screenshot(page, '15-layout-switcher');
      // The layout switcher icons should be visible above the canvas

      // ── Step 7: Test present mode ──
      console.log('\n=== Step 7: Present Mode ===');
      const presentBtn = page.locator('button:has-text("Present")').first();
      if (await presentBtn.isVisible({ timeout: 3000 }).catch(() => false)) {
        await presentBtn.click();
        await page.waitForLoadState('networkidle');
        await page.waitForTimeout(3000);
        await screenshot(page, '16-present-mode');
        console.log('  Presentation mode opened');

        // Navigate forward
        await page.keyboard.press('ArrowRight');
        await page.waitForTimeout(1000);
        await screenshot(page, '17-present-slide-2');
        console.log('  Navigated to slide 2');

        await page.keyboard.press('ArrowRight');
        await page.waitForTimeout(1000);
        await screenshot(page, '18-present-slide-3');

        // Exit presentation
        await page.keyboard.press('Escape');
        await page.waitForTimeout(1000);
      }

      // ── Step 8: Test export buttons ──
      console.log('\n=== Step 8: Export Buttons ===');
      await page.goto(`${BASE}/reports`);
      await page.waitForLoadState('networkidle');
      await page.waitForTimeout(2000);

      // Click the first report to re-open editor
      const reportCard = page.locator('[class*="cursor-pointer"][class*="rounded"]').first();
      if (await reportCard.isVisible({ timeout: 3000 }).catch(() => false)) {
        await reportCard.click();
        await page.waitForLoadState('networkidle');
        await page.waitForTimeout(2000);
        await screenshot(page, '19-editor-reopened');

        // Check PPTX and PDF buttons exist
        const pptxBtn = page.locator('button:has-text("PPTX")').first();
        const pdfBtn = page.locator('button:has-text("PDF")').first();
        console.log(`  PPTX button visible: ${await pptxBtn.isVisible().catch(() => false)}`);
        console.log(`  PDF button visible: ${await pdfBtn.isVisible().catch(() => false)}`);
        await screenshot(page, '20-export-buttons');
      }

    } else {
      console.log('  Editor canvas not visible');
      await screenshot(page, '07-no-editor');
    }

    console.log('\n=== All tests completed ===');
    console.log(`Screenshots saved to ${SCREENSHOT_DIR}/`);

  } catch (err) {
    console.error('Test error:', err.message);
    await screenshot(page, 'error-state');
  } finally {
    await browser.close();
  }
}

main().catch(console.error);

File: tests/test_reports_e2e.py
"""Playwright E2E test for the redesigned Report system."""
import os, sys

# Generate auth token before launching browser
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from ix.db.conn import conn
conn.connect()
from ix.common.security.auth import create_user_token
from sqlalchemy.orm import Session
from ix.db.models import User

with Session(conn.engine) as s:
    admin = s.query(User).filter(User.role == "admin").first()
    TOKEN = create_user_token(str(admin.email), role="admin")
    print(f"Auth token generated for {admin.email}")

from playwright.sync_api import sync_playwright

BASE = "http://localhost:3000"
SHOTS = "/tmp/report-tests"
os.makedirs(SHOTS, exist_ok=True)

def shot(page, name):
    path = f"{SHOTS}/{name}.png"
    page.screenshot(path=path)
    print(f"  [screenshot] {name}")

def main():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        ctx = browser.new_context(
            viewport={"width": 1440, "height": 900},
            color_scheme="dark",
        )
        # Inject auth cookie
        ctx.add_cookies([{
            "name": "access_token",
            "value": TOKEN,
            "domain": "localhost",
            "path": "/",
        }])
        page = ctx.new_page()
        page.set_default_timeout(15000)

        try:
            # ── Step 1: Reports page (should be authenticated) ──
            print("\n=== Step 1: Reports Page ===")
            page.goto(f"{BASE}/reports")
            page.wait_for_load_state("networkidle")
            page.wait_for_timeout(2000)
            shot(page, "01-reports-list")

            # ── Step 2: Template Picker ──
            print("\n=== Step 2: Template Picker ===")
            new_btn = page.locator('button:has-text("New Report")').first
            if new_btn.is_visible(timeout=5000):
                new_btn.click()
                page.wait_for_timeout(1000)
                shot(page, "02-template-picker")
                print("  Template picker opened")

                # Pick Macro Outlook
                macro_btn = page.locator('button:has-text("Macro Outlook")').first
                if macro_btn.is_visible(timeout=3000):
                    macro_btn.click()
                    print("  Selected Macro Outlook")
                else:
                    page.locator('button:has-text("Blank")').first.click()
                    print("  Selected Blank (fallback)")

                page.wait_for_load_state("networkidle")
                page.wait_for_timeout(3000)
                shot(page, "03-report-editor")
            else:
                print("  ERROR: New Report button not found")
                shot(page, "02-error-no-button")
                return

            # ── Step 3: Editor — verify canvas ──
            print("\n=== Step 3: Editor Canvas ===")
            canvas = page.locator('[class*="aspect-[16/9]"]').first
            if not canvas.is_visible(timeout=5000):
                print("  ERROR: Editor canvas not visible")
                shot(page, "03-error-no-canvas")
                return
            print("  Editor canvas visible")

            # Edit the title slide
            title_input = page.locator('input[placeholder*="Title"], input[placeholder*="title"]').first
            if title_input.is_visible(timeout=3000):
                title_input.fill("Q2 2026 Macro Outlook")
                print("  Title filled")
            shot(page, "04-title-filled")

            # ── Step 4: Navigate slide thumbnails ──
            print("\n=== Step 4: Slide Navigation ===")
            thumbnails = page.locator('.space-y-1\\.5 > div[class*="rounded"]')
            count = thumbnails.count()
            if count == 0:
                # Try alternate selector
                thumbnails = page.locator('[class*="cursor-pointer"][class*="border"][class*="rounded"]')
                count = thumbnails.count()
            print(f"  Found {count} slide thumbnails")

            # Visit each slide type
            for idx in range(min(count, 6)):
                thumbnails.nth(idx).click()
                page.wait_for_timeout(600)
                shot(page, f"05-slide-{idx+1}")
                print(f"  Slide {idx+1} rendered")

            # ── Step 5: Layout Picker ──
            print("\n=== Step 5: Layout Picker ===")
            new_slide_btn = page.locator('button:has-text("New Slide")').first
            if new_slide_btn.is_visible(timeout=3000):
                new_slide_btn.click()
                page.wait_for_timeout(1000)
                shot(page, "06-layout-picker")
                print("  Layout picker opened")

                # Check for layout groups
                for group in ["Structure", "Charts", "Content"]:
                    vis = page.locator(f'text="{group}"').first.is_visible(timeout=1000)
                    print(f"  Group '{group}': {'visible' if vis else 'not found'}")

                # Pick a layout (e.g. "Comparison")
                comp_btn = page.locator('button:has-text("Comparison")').first
                if comp_btn.is_visible(timeout=2000):
                    comp_btn.click()
                    page.wait_for_timeout(500)
                    shot(page, "07-comparison-slide")
                    print("  Comparison slide added")
                else:
                    # Fallback: close
                    page.keyboard.press("Escape")

            # ── Step 6: Chart Picker ──
            print("\n=== Step 6: Chart Picker ===")
            # Go to a chart slide
            thumbnails = page.locator('.space-y-1\\.5 > div[class*="rounded"]')
            count = thumbnails.count()
            if count == 0:
                thumbnails = page.locator('[class*="cursor-pointer"][class*="border"][class*="rounded"]')
                count = thumbnails.count()

            # Find a slide with "Click to add chart"
            found_chart_placeholder = False
            for idx in range(min(count, 10)):
                thumbnails.nth(idx).click()
                page.wait_for_timeout(400)
                placeholder = page.locator('button:has-text("Click to add chart")').first
                if placeholder.is_visible(timeout=500):
                    found_chart_placeholder = True
                    placeholder.click()
                    page.wait_for_timeout(2000)
                    shot(page, "08-chart-picker")
                    print("  Chart picker opened")

                    # Check search
                    search = page.locator('input[placeholder*="Search"]').first
                    print(f"  Search visible: {search.is_visible()}")

                    # Count charts
                    chart_items = page.locator('button:has(.lucide-line-chart)')
                    n = chart_items.count()
                    print(f"  Charts available: {n}")

                    if n > 0:
                        chart_items.first.click()
                        page.wait_for_timeout(2500)
                        shot(page, "09-chart-preview")
                        print("  Chart preview shown")

                        insert_btn = page.locator('button:has-text("Insert Chart")').first
                        if insert_btn.is_visible(timeout=3000):
                            insert_btn.click()
                            page.wait_for_timeout(2000)
                            shot(page, "10-chart-inserted")
                            print("  Chart inserted!")
                    else:
                        # Close
                        close = page.locator('button[aria-label="Close"]').first
                        if close.is_visible(timeout=1000):
                            close.click()
                    break

            if not found_chart_placeholder:
                print("  No chart placeholder found on any slide")

            # ── Step 7: Properties Panel ──
            print("\n=== Step 7: Properties Panel ===")
            props_btn = page.locator('button[title*="properties"]').first
            if props_btn.is_visible(timeout=2000):
                props_btn.click()
                page.wait_for_timeout(500)
                shot(page, "11-properties-panel")
                print("  Properties panel opened")

                notes = page.locator('textarea[placeholder*="Notes"]').first
                if notes.is_visible(timeout=2000):
                    notes.fill("Test speaker note for this slide")
                    print("  Speaker notes filled")
                    shot(page, "12-speaker-notes")

            # ── Step 8: Duplicate slide ──
            print("\n=== Step 8: Duplicate Slide ===")
            dup_btn = page.locator('button[title*="Duplicate"]').first
            if dup_btn.is_visible(timeout=2000):
                dup_btn.click()
                page.wait_for_timeout(500)
                shot(page, "13-after-duplicate")
                # Re-count
                thumbnails2 = page.locator('.space-y-1\\.5 > div[class*="rounded"]')
                c2 = thumbnails2.count()
                if c2 == 0:
                    thumbnails2 = page.locator('[class*="cursor-pointer"][class*="border"][class*="rounded"]')
                    c2 = thumbnails2.count()
                print(f"  Slides after duplicate: {c2}")

            # ── Step 9: Export Buttons ──
            print("\n=== Step 9: Export Buttons ===")
            pptx = page.locator('button:has-text("PPTX")').first
            pdf = page.locator('button:has-text("PDF")').first
            present = page.locator('button:has-text("Present")').first
            print(f"  PPTX visible: {pptx.is_visible()}")
            print(f"  PDF visible: {pdf.is_visible()}")
            print(f"  Present visible: {present.is_visible()}")
            shot(page, "14-toolbar")

            # ── Step 10: Present Mode ──
            print("\n=== Step 10: Present Mode ===")
            if present.is_visible():
                present.click()
                page.wait_for_load_state("networkidle")
                page.wait_for_timeout(3000)
                shot(page, "15-present-slide-1")
                print("  Presentation mode active")

                for i in range(4):
                    page.keyboard.press("ArrowRight")
                    page.wait_for_timeout(800)
                    shot(page, f"16-present-slide-{i+2}")
                print("  Navigated through slides")

                page.keyboard.press("Escape")
                page.wait_for_timeout(1500)
                shot(page, "17-back-from-present")
                print("  Exited presentation")

            print("\n=============================")
            print("=== ALL TESTS COMPLETE ===")
            print("=============================")
            print(f"Screenshots at: {SHOTS}/")

        except Exception as e:
            print(f"\nERROR: {e}")
            import traceback; traceback.print_exc()
            shot(page, "error-state")
        finally:
            browser.close()

if __name__ == "__main__":
    main()
